### PR TITLE
Free Text Search Filter (DS-723)

### PR DESCRIPTION
**Pull request by ejgullo**
Free Text Search Filter (DS-723)
- Renamed files to match properly (previous mix of ‘keyword’ and ‘search’ - all are now keyword to avoid any confusion w/ overall search)
- Removed ‘coming soon’ inactive state
- Added submit button, hooked up to API to put keyword(s) into redux and pass into search terms (Searches against the ‘description’ field only)
- Created query builder + updated searchOperation to include the filter
- Updated top filter bar to include search term tag (leaving this as an array, both to work with the base filter group generator and to allow for future updates that are likely to include AND/OR search operations and result in multiple terms)
- Searching simply the exact string in this sprint, further updates to the behavior are out of scope for the current sprint

### CODE REVIEW

File: src/js/containers/search/filters/KeywordContainer.jsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        }
+        // take in keywords and pass to redux
+        if (this.state.newKeyword !== null) {
+            this.props.updateTextSearchInput(this.state.newKeyword);
+        }
+    }
+
+    // add redux value to filter values
+
+    // resubmit filters
```
Comment (mab6bf) [inline src/js/containers/search/filters/KeywordContainer.jsx]: Are these old comments?

File: src/js/components/search/filters/keyword/Keyword.jsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    render() {
+        return (
+            <div className="keyword-filter search-filter">
+                <form
+                    onSubmit={this.props.submitText}>
+                    <input
+                        id="search"
+                        type="text"
+                        className="keyword-input"
+                        placeholder="eg: Education"
```
Comment (mab6bf) [inline src/js/components/search/filters/keyword/Keyword.jsx]: Let's confirm what UX wants here, but it should at least be "e.g., Education"

File: tests/redux/reducers/search/searchFiltersReducer-test.js [hunk 3]
```diff
@@ -133,6 +134,18 @@ describe('searchFiltersReducer', () => {
         });
     });
 
+    describe('UPDATE_TEXT_SEARCH', () => {
```
Comment (mab6bf) [inline tests/redux/reducers/search/searchFiltersReducer-test.js]: This is a good start, but we'll also need to add tests to `TopFilterBarContainer-test` (showing filter when present, removing from redux when X'd out), create `KeywordContainer-test` tests, and also update the `defaultFilters` object in `testResources/defaultReduxFilters`.

File: src/js/redux/reducers/search/searchFiltersReducer.js [hunk 4]
```diff
@@ -99,6 +100,11 @@ const searchFiltersReducer = (state = initialState, action) => {
                         state.selectedRecipientLocations, action.location)
             });
         }
+        case 'UPDATE_TEXT_SEARCH': {
```
Comment (mab6bf) [inline src/js/redux/reducers/search/searchFiltersReducer.js]: Can you move this to the top of the `switch` statement, adding an appropriate comment above, so this statement can match the order of the Sidebar filters?

File: src/js/redux/actions/search/searchFilterActions.js [hunk 5]
```diff
@@ -82,6 +82,12 @@ export const updateRecipientLocations = (state) => ({
     location: state
 });
 
+// Keyword Filter
```
Comment (mab6bf) [inline src/js/redux/actions/search/searchFilterActions.js]: Same comment here - can you move this to the top of the file to be more in order with the Sidebar filters?

File: src/js/containers/search/topFilterBar/TopFilterBarContainer.jsx [hunk 6]
```diff
@@ -140,6 +146,30 @@ export class TopFilterBarContainer extends React.Component {
...
+        let selected = false;
+        const filter = {};
+
+        if (props.keyword) {
+        // keyword exists
+            selected = true;
+            filter.code = 'keyword';
+            filter.name = 'Keyword';
+
+            filter.values = props.keyword;
```
Comment (mab6bf) [inline src/js/containers/search/topFilterBar/TopFilterBarContainer.jsx]: We need to figure out how to handle different filter type lengths in `TopFilterBar`. I started this fix in Award Amount, but it applies here too, to String types; the "length" of this filter is technically the length of the keyword string, so typing in "Education" will show "9 Current Filters".

File: src/js/components/search/topFilterBar/filterGroups/KeywordFilterGroup.jsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    generateTags() {
+        const tags = [];
+
+        // check to see if a keyword is provided
+        const keyword = this.props.filter.values;
+
+        const tag = {
+            value: `${keyword}`,
```
Comment (mab6bf) [inline src/js/components/search/topFilterBar/filterGroups/KeywordFilterGroup.jsx]: We shouldn't need this in a template literal

File: src/js/components/search/topFilterBar/filterGroups/TopFilterGroupGenerator.jsx [hunk 8]
```diff
@@ -40,6 +41,8 @@ export const topFilterGroupGenerator = (config = {
             return <AgencyFilterGroup key={groupKey} {...config} />;
         case 'selectedRecipients':
             return <RecipientFilterGroup key={groupKey} {...config} />;
+        case 'keyword':
```
Comment (mab6bf) [inline src/js/components/search/topFilterBar/filterGroups/TopFilterGroupGenerator.jsx]: Same comment as in others - can we move this to the top to reflect the order of the filters in the Sidebar?

File: src/js/models/search/SearchOperation.js [hunk 9]
```diff
@@ -103,6 +107,11 @@ class SearchOperation {
             );
         }
 
+        // // add keyword query
```
Comment (mab6bf) [inline src/js/models/search/SearchOperation.js]: To the top!

File: src/js/models/search/SearchOperation.js [hunk 10]
```diff
@@ -19,6 +20,8 @@ class SearchOperation {
         this.timePeriodFY = [];
         this.timePeriodRange = [];
 
+        this.keyword = '';
```
Comment (mab6bf) [inline src/js/models/search/SearchOperation.js]: To the top!

File: src/js/models/search/SearchOperation.js [hunk 11]
```diff
@@ -45,6 +48,7 @@ class SearchOperation {
             this.timePeriodRange = [state.timePeriodStart, state.timePeriodEnd];
             this.timePeriodFY = [];
         }
+        this.keyword = state.keyword;
```
Comment (mab6bf) [inline src/js/models/search/SearchOperation.js]: To the top!

File: src/js/containers/search/filters/KeywordContainer.jsx [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+    handleChange(e) {
+        this.setState({
+            newKeyword: e.target.value
+        });
+    }
+
+    submitText() {
+        // remove any previous keyword applied in redux
+        if (this.props.keyword !== null) {
```
Comment (mab6bf) [inline src/js/containers/search/filters/KeywordContainer.jsx]: This first `if` statement is unnecessary. Redux, as a state machine, will simply overwrite the old keyword with the new keyword, so there is no reason to clear the filter type.

File: src/js/containers/search/filters/KeywordContainer.jsx [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        this.state = {
+            newKeyword: '',
+            keyword: ''
+        };
+
+        this.submitText = this.submitText.bind(this);
+        this.handleChange = this.handleChange.bind(this);
+    }
+
+    handleChange(e) {
```
Comment (mab6bf) [inline src/js/containers/search/filters/KeywordContainer.jsx]: We shouldn't need two methods to grab text from the user input. I'd add a `ref` on the text input in `Keyword`, as well as a function that grabs the input from the `ref` when the submit button is clicked. That function will call `this.props.submitText` with the user-input value from the `ref` input, which will remove the need for both `handleChange` and keeping track of input in `state`.

Comment (ejgullo) [inline src/js/containers/search/filters/KeywordContainer.jsx]: 👍   Thanks for the suggestion here, cleared up a lot of unnecessary code and helped me get a better understanding of how `ref` functions!

File: src/js/containers/search/filters/KeywordContainer.jsx (hunk 12 above)
Comment (ejgullo) [inline src/js/containers/search/filters/KeywordContainer.jsx]: Fixed via the above fix~

File: src/js/containers/search/filters/KeywordContainer.jsx (hunk 1 above)
Comment (ejgullo) [inline src/js/containers/search/filters/KeywordContainer.jsx]: Whoops.

File: src/js/components/search/filters/keyword/Keyword.jsx (hunk 2 above)
Comment (ejgullo) [inline src/js/components/search/filters/keyword/Keyword.jsx]: Confirming now!

File: src/js/redux/reducers/search/searchFiltersReducer.js (hunk 4 above)
Comment (ejgullo) [inline src/js/redux/reducers/search/searchFiltersReducer.js]: Will do!

File: tests/redux/reducers/search/searchFiltersReducer-test.js (hunk 3 above)
Comment (ejgullo) [inline tests/redux/reducers/search/searchFiltersReducer-test.js]: 👍

File: src/js/redux/actions/search/searchFilterActions.js (hunk 5 above)
Comment (ejgullo) [inline src/js/redux/actions/search/searchFilterActions.js]: 👍

File: src/js/components/search/topFilterBar/filterGroups/KeywordFilterGroup.jsx (hunk 7 above)
Comment (ejgullo) [inline src/js/components/search/topFilterBar/filterGroups/KeywordFilterGroup.jsx]: 👍

File: src/js/containers/search/topFilterBar/TopFilterBarContainer.jsx (hunk 6 above)
Comment (ejgullo) [inline src/js/containers/search/topFilterBar/TopFilterBarContainer.jsx]: @mab6bf Where do we call for the length of the filter?

File: src/js/components/search/topFilterBar/filterGroups/TopFilterGroupGenerator.jsx (hunk 8 above)
Comment (ejgullo) [inline src/js/components/search/topFilterBar/filterGroups/TopFilterGroupGenerator.jsx]: 👍

File: src/js/models/search/SearchOperation.js (hunk 10 above)
Comment (ejgullo) [inline src/js/models/search/SearchOperation.js]: 👍

File: src/js/models/search/SearchOperation.js (hunk 11 above)
Comment (ejgullo) [inline src/js/models/search/SearchOperation.js]: 👍

File: src/js/models/search/SearchOperation.js (hunk 9 above)
Comment (ejgullo) [inline src/js/models/search/SearchOperation.js]: 👍

File: src/js/containers/search/topFilterBar/TopFilterBarContainer.jsx (hunk 6 above)
Comment (mab6bf) [inline src/js/containers/search/topFilterBar/TopFilterBarContainer.jsx]: TopFilterBar.jsx, line 40: ```const filters = this.props.filters.map((filter) => {
            filterCount += filter.values.length;``` // if this is a string, it'll add the length of the string
...
}
```

File: src/js/components/search/filters/keyword/Keyword.jsx [hunk 14]
```diff
@@ -6,27 +6,32 @@
...
+    searchKeyword() {
+        this.props.submitText(this.keyword.value);
+    }
+
     render() {
         return (
             <div className="keyword-filter search-filter">
-                <form
-                    onSubmit={this.props.submitText}>
+                <form>
```
Comment (mab6bf) [inline src/js/components/search/filters/keyword/Keyword.jsx]: We shouldn't need to wrap this in a form

File: tests/containers/search/keyword/KeywordContainer-test.jsx [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        it('should submit given keyword text to redux state', () => {
+            const mockReduxActionKeyword = jest.fn((args) => {
+                expect(args).toEqual('Education');
+            });
+            const keywordContainer = shallow(
+                <KeywordContainer
+                    keyword={initialFilters}
+                    updateTextSearchInput={mockReduxActionKeyword} />);
+
+            const toggleAgencySpy = sinon.spy(keywordContainer.instance(),
```
Comment (mab6bf) [inline tests/containers/search/keyword/KeywordContainer-test.jsx]: Can we de-Agency-Filter-fy this file by naming this `submitTextSpy`?

File: src/js/components/search/filters/keyword/Keyword.jsx (hunk 14 above)
Comment (ejgullo) [inline src/js/components/search/filters/keyword/Keyword.jsx]: 👍

File: tests/containers/search/keyword/KeywordContainer-test.jsx (hunk 15 above)
Comment (ejgullo) [inline tests/containers/search/keyword/KeywordContainer-test.jsx]: Whoops. 😅

**Pull request merged**